pkg/tfbridge/rawstate.go

```diff
@@ -0,0 +1,768 @@
+			infl, err := ih.inflectionsAt(subPath, pvElements[key], e)
+			if err != nil {
+				return rawStateInflections{}, err
+			}
+			mapInfl.set(key, infl)
+		}
+
+		return rawStateInflections{Map: &mapInfl}, nil
+
+	case v.Type().IsSetType():
+		// Key assumption here is that when Pulumi translates Set values in states and projects them as Array
+		// PropertyValue, the Array preserves the original ordering.
+
+		elements := v.AsValueSlice()
+
+		// MaxItems=1 handling is exactly similar to lists.
+		//
+		// Checking if [] got encoded as Null due to MaxItems=1.
+		if len(elements) == 0 && pv.IsNull() {
+			t := v.Type().ElementType()
+			return rawStateInflections{
+				Pluralize: &pluralize{
+					ElementType: &t,
+					IsSet:       true,
+				},
+			}, nil
+		}
+
+		// Checking if [x] got encoded as x due to MaxItems=1.
+		if len(elements) == 1 && !pv.IsArray() {
+			subPath := path.Element()
+			inner, err := ih.inflectionsAt(subPath, pv, elements[0])
+			if err != nil {
+				return rawStateInflections{}, err
+			}
+			return rawStateInflections{Pluralize: &pluralize{
+				Inner: inner,
+				IsSet: true,
+			}}, nil
+		}
+
+		// Otherwise PropertyValue should be an array just like the cty.Value is a set.
+		contract.Assertf(pv.IsArray(), "Expected an Array PropertyValue to match a Set cty.Value")
+
+		pvElements := pv.ArrayValue()
+
+		contract.Assertf(len(pvElements) == len(elements),
+			"Expected array length parity for PropertyValue and matching Set cty.Value")
+
+		if len(pvElements) == 0 {
+			return rawStateInflections{
+				Set: &setInflections{T: v.Type().SetElementType()},
+			}, nil
+		}
+
+		setInfl := setInflections{}
+
+		subPath := path.Element()
+		for k, e := range elements {
+			infl, err := ih.inflectionsAt(subPath, pvElements[k], e)
+			if err != nil {
+				return rawStateInflections{}, err
+			}
+			setInfl.set(k, infl)
+		}
+
+		return rawStateInflections{Set: &setInfl}, nil
+
+	case v.Type().IsObjectType():
+		elements := v.AsValueMap()
+		pvElements := pv.ObjectValue()
+
+		if len(pvElements) == 0 {
+			infl := objInflections{}
+			for k := range pvElements {
+				infl.ignore(k)
+			}
+			return rawStateInflections{Obj: &infl}, nil
+		}
+
+		infl := objInflections{}
+
+		keySetWithCtyValueMatches := map[resource.PropertyKey]struct{}{}
+
+		for k, v := range elements {
+			subPath := path.GetAttr(k)
+			keyRaw, err := TerraformToPulumiNameAtPath(subPath, ih.schemaMap, ih.schemaInfos)
+			if err != nil {
+				return rawStateInflections{}, err
+			}
+			key := resource.PropertyKey(keyRaw)
+			kInfl, err := ih.inflectionsAt(subPath, pvElements[key], v)
+			if err != nil {
+				return rawStateInflections{}, err
+			}
+			keySetWithCtyValueMatches[key] = struct{}{}
+			infl.set(k, key, kInfl)
+		}
+
+		for k := range pvElements {
+			if _, ok := keySetWithCtyValueMatches[k]; !ok {
+				infl.ignore(k)
+			}
+		}
+
+		return rawStateInflections{Obj: &infl}, nil
+	case v.Type().IsTupleType():
+		panic("TODO TypeType")
```

Comment:
Tuples should be possible in principle in resource states? Perhaps wrapped under DynamicValue ?

 https://developer.hashicorp.com/terraform/plugin/framework/handling-data/types/tuple says it is limited.  Curious there are no bridge tests exercising these.